source/common/router/router_ratelimit.cc

```diff
@@ -109,28 +109,31 @@ void RateLimitPolicyEntryImpl::populateDescriptors(
 
 RateLimitPolicyImpl::RateLimitPolicyImpl(const Json::Object& config) {
   if (config.hasObject("rate_limits")) {
-    std::vector<std::unique_ptr<RateLimitPolicyEntry>> rate_limit_policy;
-    std::vector<std::reference_wrapper<const RateLimitPolicyEntry>> rate_limit_policy_reference;
+    // The supported number of stages is 0 through RateLimitPolicyImpl::MAX_STAGE_NUMBER
+    // inclusively.
+    for (size_t i = 0; i <= RateLimitPolicyImpl::MAX_STAGE_NUMBER; i++) {
+      std::vector<std::unique_ptr<RateLimitPolicyEntry>> rate_limit_policy;
+      std::vector<std::reference_wrapper<const RateLimitPolicyEntry>> rate_limit_policy_reference;
+      rate_limit_entries_reference_.emplace_back(rate_limit_policy_reference);
+      rate_limit_entries_.emplace_back(std::move(rate_limit_policy));
+    }
+
     for (const Json::ObjectPtr& rate_limit : config.getObjectArray("rate_limits")) {
       std::unique_ptr<RateLimitPolicyEntry> rate_limit_policy_entry(
           new RateLimitPolicyEntryImpl(*rate_limit));
-      rate_limit_policy_reference.emplace_back(*rate_limit_policy_entry);
-      rate_limit_policy.emplace_back(std::move(rate_limit_policy_entry));
+      uint64_t stage = rate_limit_policy_entry->stage();
+      rate_limit_entries_reference_[stage].emplace_back(*rate_limit_policy_entry);
+      rate_limit_entries_[stage].emplace_back(std::move(rate_limit_policy_entry));
     }
-    rate_limit_entries_.emplace_back(std::move(rate_limit_policy));
-    rate_limit_entries_reference_.emplace_back(rate_limit_policy_reference);
   }
 }
 
-const std::vector<std::reference_wrapper<const RateLimitPolicyEntry>>&
-    RateLimitPolicyImpl::getApplicableRateLimit(int64_t) const {
-  // Currently return all rate limit policy entries.
-  // TODO(mattklein123): Implement returning only rate limit policy entries that match the stage
-  // setting.
-  if (rate_limit_entries_.empty()) {
+const std::vector<std::reference_wrapper<const Router::RateLimitPolicyEntry>>&
+RateLimitPolicyImpl::getApplicableRateLimit(uint64_t stage) const {
+  if (rate_limit_entries_.empty() || stage > RateLimitPolicyImpl::MAX_STAGE_NUMBER) {
```

Comment:
stage > RateLimitPolicyImpl::MAX_STAGE_NUMBER should never happen because of schema. If you want to keep the check that's fine, but add an ASSERT.